distributed-job-manager/src/main/kotlin/ru/fix/distributed/job/manager/Rebalancer.kt

```diff
@@ -0,0 +1,261 @@
+package ru.fix.distributed.job.manager
+
+import org.apache.curator.framework.CuratorFramework
+import org.apache.curator.framework.imps.CuratorFrameworkState
+import org.apache.logging.log4j.kotlin.Logging
+import org.apache.zookeeper.KeeperException.NoNodeException
+import ru.fix.aggregating.profiler.Profiler
+import ru.fix.distributed.job.manager.model.AssignmentState
+import ru.fix.distributed.job.manager.model.JobId
+import ru.fix.distributed.job.manager.model.WorkItem
+import ru.fix.distributed.job.manager.model.WorkerId
+import ru.fix.distributed.job.manager.strategy.AssignmentStrategy
+import ru.fix.stdlib.concurrency.events.ReducingEventAccumulator
+import ru.fix.stdlib.concurrency.threads.NamedExecutors
+import ru.fix.zookeeper.transactional.ZkTransaction
+import ru.fix.zookeeper.utils.ZkTreePrinter
+import java.util.*
+import java.util.concurrent.TimeUnit
+
+private const val ASSIGNMENT_COMMIT_RETRIES_COUNT = 3
+
+private val REBALANCE_EVENT = Any()
+
+internal class Rebalancer(
+        profiler: Profiler,
+        private val paths: ZkPathsManager,
+        private val curatorFramework: CuratorFramework,
+        private val leaderLatchExecutor: LeaderLatchExecutor,
+        private val assignmentStrategy: AssignmentStrategy,
+        private val nodeId: String
+) : AutoCloseable {
+    private val zkPrinter = ZkTreePrinter(curatorFramework)
+
+    private val rebalanceEventReceiverExecutor = NamedExecutors.newSingleThreadPool(
+            "rebalance-event-receiver", profiler)
+
+    private val rebalanceEventAccumulator = ReducingEventAccumulator.lastEventWinAccumulator<Any>()
+
+    fun handleRebalanceEvent() {
+        rebalanceEventAccumulator.publishEvent(REBALANCE_EVENT)
+    }
+
+    fun start() = rebalanceEventReceiverExecutor.execute {
```

Comment:
method names are not clear and this probably confuse the situation little bit:
leaderLatchExecutor.submitIfNeeded, what this `needed` means? Actually it is `submitIfHasLeadership` but what if we submit task, this task will be putted into executor queue, then we  lose leadership and then executor take task from the queue and try to execute it?   Probably we should  wrap runnable and execute it only if we still have leadership after queuing task.

Reply:
- method submitIfNeeded renamed to submitIfHasLeadership

- task wrapped for double check leadership in LeaderLatchExecutor

- leadership check added into rebalance TransactionCreator to reduce tries to commit rebalance by non-leader manager

Reply:
Now i think LeaderLatchExecutor was my mistake. Its executor really needed only for rebalancing. So I threw away internal executor and renamed LeaderLatchExecutor to ManagerState.
Also i made ManagerState receive all events from Manager, so Rebalancer is not responsible for that anymore.
I updated PR description, please take a look.